pkg/aws/client.go

```diff
@@ -293,29 +291,72 @@ func (b *ClientBuilder) UseLocalCredentials(value bool) *ClientBuilder {
 }
 
 // Create AWS session with a specific set of credentials
-func (b *ClientBuilder) BuildSessionWithOptionsCredentials(value *AccessKey) (*session.Session, error) {
-	return session.NewSessionWithOptions(session.Options{
-		Config: aws.Config{
-			CredentialsChainVerboseErrors: aws.Bool(true),
-			Region:                        b.region,
-			Credentials: credentials.NewStaticCredentials(
-				value.AccessKeyID,
-				value.SecretAccessKey,
-				"",
-			),
-		},
-	})
+func (b *ClientBuilder) BuildSessionWithOptionsCredentials(value *AccessKey,
+	logLevel aws.ClientLogMode) (*aws.Config, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO(),
+		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(value.AccessKeyID,
+			value.SecretAccessKey, "")),
+		config.WithRegion(*b.region),
+		config.WithHTTPClient(&http.Client{
+			Transport: http.DefaultTransport,
+		}),
+		config.WithClientLogMode(logLevel),
+		config.WithRetryer(func() aws.Retryer {
+			retryer := retry.AddWithMaxAttempts(retry.NewStandard(), NumMaxRetries)
+			retryer = retry.AddWithMaxBackoffDelay(retryer, time.Second)
+			retryer = retry.AddWithErrorCodes(retryer, awserr.InvalidClientTokenID)
+			return retryer
+		}),
+		config.WithAPIOptions([]func(stack *middleware.Stack) error{
+			smithyhttp.AddHeaderValue("User-Agent",
+				strings.Join([]string{"ROSACLI", info.Version}, ";")),
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cfg, nil
```

Comment:
`BuildSession` triggers both `BuildSessionWithOptionsCredentials` and `BuildSessionWithOptions`, we return `*aws.config`, then inside `Build()` we reference the pointer when building the aws client:
```
	c := &awsClient{
		cfg:                 *cfg,
```
It would be better to return `aws.config` instead of the back and forth, to improve readability and reduce pointers, WDYT?

Reply:
@oriAdler I think `BuildSession` only triggers one or the other as it returns the result of the function in the `if` statement e.g:

```
if b.credentials != nil {
		return b.BuildSessionWithOptionsCredentials(b.credentials, logLevel)
	}
```

Am I correct?

Reply:
yes it only returns one or the other based on the if statement

Reply:
@robpblake yes, thanks for the clarification.
We can still change the return value of the function to `aws.config`, to simplify the code.